cli/dcos-spark/security.go

```diff
@@ -0,0 +1,236 @@
+		paths = append(paths, args.truststoreSecretPath)
+		filenames = append(filenames, trustStoreFileName)
+		envkeys = append(envkeys, "DCOS_SPARK_TRUSTSTORE")
+	}
+	joinedPaths := strings.Join(paths, ",")
+	joinedFilenames := strings.Join(filenames, ",")
+	joinedEnvkeys := strings.Join(envkeys, ",")
+
+	taskTypes := []string{"driver", "executor"}
+	for _, taskType := range taskTypes {
+		appendToProperty(fmt.Sprintf(SECRET_REFERENCE_TEMPLATE, taskType), joinedPaths, args)
+		appendToProperty(fmt.Sprintf(SECRET_FILENAME_TEMPLATE, taskType), joinedFilenames, args)
+		appendToPropertyIfSet(fmt.Sprintf(SECRET_ENVKEY_TEMPLATE, taskType), joinedEnvkeys, args)
+	}
+
+	// Passwords
+	args.properties["spark.ssl.keyStorePassword"] = args.keystorePassword
+	args.properties["spark.ssl.keyPassword"] = args.privateKeyPassword
+
+	if args.truststoreSecretPath != "" {
+		args.properties["spark.ssl.trustStorePassword"] = args.truststorePassword
+	}
+
+	// Protocol
+	if _, ok := args.properties["spark.ssl.protocol"]; !ok {
+		args.properties["spark.ssl.protocol"] = "TLS"
+	}
+}
+
+func setupSaslProperties(secretPath string, args *sparkArgs) {
+	args.properties["spark.mesos.containerizer"] = "mesos"
+	args.properties["spark.authenticate"] = "true"
+	args.properties["spark.authenticate.enableSaslEncryption"] = "true"
+	args.properties["spark.authenticate.secret"] = "spark_shared_secret"
+	args.properties["spark.executorEnv._SPARK_AUTH_SECRET"] = "spark_shared_secret"
+	taskTypes := []string{"driver", "executor"}
+	for _, taskType := range taskTypes {
+		secretRefProp := fmt.Sprintf(SECRET_REFERENCE_TEMPLATE, taskType)
+		secretFileProp := fmt.Sprintf(SECRET_FILENAME_TEMPLATE, taskType)
+		appendToProperty(secretRefProp, secretPath, args)
+		appendToProperty(secretFileProp, prepareBase64Secret(secretPath), args)
+	}
+}
+
+func generateRandomBytes(n int) ([]byte, error) {
+	// https://elithrar.github.io/article/generating-secure-random-numbers-crypto-rand/
+	b := make([]byte, n)
+	_, err := rand.Read(b)
+	if err != nil {
+		return nil, err
+	}
+	return b, nil
+}
+
+func GetRandomStringSecret() (string, error) {
+	b, err := generateRandomBytes(KEYLENGTH)
+	return base64.URLEncoding.EncodeToString(b), err
+}
+
+// KERBEROS
+
+func ValidateInputs(args *sparkArgs) error {
+	if args.keytabSecretPath == "" && args.tgtSecretPath == "" && args.tgtSecretValue == "" {
+		return errors.New("need to provide Keytab secret, TGT secret, or TGT value " +
+			"with Kerberos principal")
+	}
+
+	if args.keytabSecretPath != "" && (args.tgtSecretValue != "" || args.tgtSecretPath != "") {
+		return errors.New("keytabs and TGTs cannot be used together")
+	}
+
+	if args.tgtSecretPath != "" && args.tgtSecretValue != "" {
+		return errors.New("cannot use a TGT-by-value and a TGT-by-secret at the same time")
+	}
+	return nil
+}
+
+func GetPrincipal(args *sparkArgs) string {
+	principal, contains := args.properties["spark.yarn.principal"]
+	if contains {
+		return principal
```

Comment:
Should this not check for a non-empty principal?

Reply:
If the principal is empty then we won't start down the kerberos fork:
```go
// kerberos configuration:
	kerberosPrincipal := GetPrincipal(args)
	if kerberosPrincipal != "" {
		err := ValidateInputs(args)
		if err != nil {
			return "", err
		}
```